cluster/agent.go

```diff
@@ -0,0 +1,259 @@
+	case *ActivationRequest:
+		resp := a.handleActivationRequest(msg)
+		c.Respond(resp)
+	case getMembers:
+		c.Respond(a.members.Slice())
+	case getKinds:
+		kinds := make([]string, len(a.kinds))
+		i := 0
+		for kind := range a.kinds {
+			kinds[i] = kind
+			i++
+		}
+		c.Respond(kinds)
+	case getActive:
+		pid := a.activated[msg.id]
+		c.Respond(pid)
+	}
+}
+
+func (a *Agent) handleActorTopology(msg *ActorTopology) {
+	for _, actorInfo := range msg.Actors {
+		a.addActivated(actorInfo.PID)
+	}
+}
+
+func (a *Agent) handleDeactivation(msg *Deactivation) {
+	a.removeActivated(msg.PID)
+	a.cluster.engine.Poison(msg.PID)
+	a.cluster.engine.BroadcastEvent(DeactivationEvent{PID: msg.PID})
+}
+
+// A new kind is activated on this cluster.
+func (a *Agent) handleActivation(msg *Activation) {
+	a.addActivated(msg.PID)
+	a.cluster.engine.BroadcastEvent(ActivationEvent{PID: msg.PID})
+}
+
+func (a *Agent) handleActivationRequest(msg *ActivationRequest) *ActivationResponse {
+	if !a.hasKindLocal(msg.Kind) {
+		slog.Error("received activation request but kind not registered locally on this node", "kind", msg.Kind)
+		return &ActivationResponse{Success: false}
+	}
+	kind := a.localKinds[msg.Kind]
+	pid := a.cluster.engine.Spawn(kind.producer, msg.Kind+"/"+msg.ID)
+	resp := &ActivationResponse{
+		PID:     pid,
+		Success: true,
+	}
+	return resp
+}
+
+func (a *Agent) activate(kind, id string) *actor.PID {
+	members := a.members.FilterByKind(kind)
+	owner := a.cluster.activationStrategy.SelectMember(members)
+	if owner == nil {
+		panic("TODO")
+	}
+	req := &ActivationRequest{Kind: kind, ID: id}
+	activatorPID := actor.NewPID(owner.Host, "cluster/"+owner.ID)
+
+	var activationResp *ActivationResponse
+	// Local activation
+	if owner.Host == a.cluster.engine.Address() {
+		activationResp = a.handleActivationRequest(req)
+	} else {
+		// Remote activation
+
+		// TODO: topology hash
+		resp, err := a.cluster.engine.Request(activatorPID, req, requestTimeout).Result()
+		if err != nil {
+			slog.Error("failed activation request", "err", err)
+			return nil
+		}
+		r, ok := resp.(*ActivationResponse)
+		if !ok {
+			slog.Error("expected *ActivationResponse", "msg", reflect.TypeOf(resp))
+			return nil
+		}
+		if !r.Success {
+			slog.Error("activation unsuccessfull", "msg", r)
+			return nil
+		}
+		activationResp = r
+	}
+
+	a.bcast(&Activation{
+		PID: activationResp.PID,
+	})
+
+	return activationResp.PID
+}
+
+func (a *Agent) handleMembers(members []*Member) {
+	joined := NewMemberSet(members...).Except(a.members.Slice())
+	left := a.members.Except(members)
+
+	for _, member := range joined {
+		a.memberJoin(member)
+	}
+	for _, member := range left {
+		a.memberLeave(member)
+	}
+}
+
+func (a *Agent) memberJoin(member *Member) {
+	a.members.Add(member)
+
+	// track cluster wide available kinds
+	for _, kind := range member.Kinds {
+		if _, ok := a.kinds[kind]; !ok {
+			a.kinds[kind] = true
+		}
+	}
+
+	actorInfos := []*ActorInfo{}
```

Comment:
Maybe this is a style thing, but is there a good reason not to just make([]*ActorInfo,0) here?

Reply:
Well, it's the same. More readable imo, no?

Reply:
Personally, I don't think so. Maybe that is just preference, but I've been tought to use make(), so that is what I'm used to. :-)

 I think a nil declaration will work here, though. So just a `var actorInfos []*ActorInfo` should work and is somewhat prettier.